exercises/forall-param-reductions_solution.cpp

```diff
@@ -0,0 +1,293 @@
+
+/*
+ *  Reduction Example
+ *
+ *  This example illustrates use of the RAJA reduction types: min, max,
+ *  sum, min-loc, and max-loc.
+ *
+ *  RAJA features shown:
+ *    - `forall` loop iteration template method
+ *    -  Index range segment
+ *    -  Execution policies
+ *    -  Reduction types
+ *
+ * If CUDA is enabled, CUDA unified memory is used.
+ */
+
+/*
+  Specify the number of threads in a GPU thread block
+*/
+#if defined(RAJA_ENABLE_CUDA)
+constexpr int CUDA_BLOCK_SIZE = 256;
+#endif
+
+#if defined(RAJA_ENABLE_HIP)
+constexpr int HIP_BLOCK_SIZE = 256;
+#endif
+
+int main(int RAJA_UNUSED_ARG(argc), char** RAJA_UNUSED_ARG(argv[]))
+{
+
+  std::cout << "\n\nRAJA reductions example...\n";
+
+  // _reductions_array_init_start
+//
+// Define array length
+//
+  constexpr int N = 1000000;
+
+//
+// Allocate array data and initialize data to alternating sequence of 1, -1.
+//
+  int* a = memoryManager::allocate<int>(N);
+
+  for (int i = 0; i < N; ++i) {
+    if ( i % 2 == 0 ) {
+      a[i] = 1;
+    } else {
+      a[i] = -1; 
+    }
+  }
+
+//
+// Set min and max loc values
+//
+  constexpr int minloc_ref = N / 2;
+  a[minloc_ref] = -100;
+
+  constexpr int maxloc_ref = N / 2 + 1;
+  a[maxloc_ref] = 100;
+  // _reductions_array_init_end
+
+//
+// Note: with this data initialization scheme, the following results will
+//       be observed for all reduction kernels below:
+//
+//  - the sum will be zero
+//  - the min will be -100
+//  - the max will be 100
+//  - the min loc will be N/2
+//  - the max loc will be N/2 + 1
+//
+//
+
+//
+// Define index range for iterating over a elements in all examples
+//
+  // _reductions_range_start
+  RAJA::TypedRangeSegment<int> arange(0, N);
+  // _reductions_range_end
+
+//
+// Define ValLoc Type
+//
+
+  using VALLOC_INT = RAJA::expt::ValLoc<int>;
+//----------------------------------------------------------------------------//
+
+  std::cout << "\n Running RAJA sequential reductions...\n";
+
+  // _reductions_raja_seq_start
+  using EXEC_POL1   = RAJA::seq_exec;
+  using REDUCE_POL1 = RAJA::seq_reduce;
+ 
+  int seq_sum = 0;
+  int seq_min = std::numeric_limits<int>::max();
+  int seq_max = std::numeric_limits<int>::min();
+  VALLOC_INT seq_minloc(std::numeric_limits<int>::max(), -1);
+  VALLOC_INT seq_maxloc(std::numeric_limits<int>::min(), -1);
+
+  RAJA::forall<EXEC_POL1>(arange, 
+    RAJA::expt::Reduce<RAJA::operators::plus>(&seq_sum),
+    RAJA::expt::Reduce<RAJA::operators::minimum>(&seq_min),
+    RAJA::expt::Reduce<RAJA::operators::maximum>(&seq_max),
+    RAJA::expt::Reduce<RAJA::operators::minimum>(&seq_minloc),
+    RAJA::expt::Reduce<RAJA::operators::maximum>(&seq_maxloc),
+    [=](int i, int &_seq_sum, int &_seq_min, int &_seq_max, VALLOC_INT &_seq_minloc, VALLOC_INT &_seq_maxloc) {
+      _seq_sum += a[i];
+
+      _seq_min = RAJA_MIN(a[i], _seq_min);
+      _seq_max = RAJA_MAX(a[i], _seq_max);
+
+      _seq_minloc = RAJA_MIN(VALLOC_INT(a[i], i), _seq_minloc);
```

Comment:
Do we want to promote the macro usage for 'loc' reductions or the method call (below)? The method calls are more compact, but the macro is more consistent with the non-loc min/max reduction usage.

Reply:
I'm not sure, I think a consistent interface is cleaner, but I wanted to continue to provide the current functionality where I could

Reply:
I'm not sure either, but the macro version makes a temporary VALLOC_INT object so it could be more expensive, right?  Or do compilers optimize that out since it's not named, similar to RVO?

Reply:
That sounds like a question for compiler explorer

Reply:
https://godbolt.org/z/E191Gc18M : It looks like `.min()` is faster only with `-O0`, `-O1` and above optimizes away from what I can tell.

Reply:
Makes sense. Thanks for checking.